### PR TITLE
[WIP] Create pandas_interoperability

### PR DESCRIPTION
**Pull request by jamartinh**
A collection of functions for inter-operability between Orange Tables and Pandas Data Frame.

### CODE REVIEW

**Comment by kernc** [comment]
Thanks for the pull request!

Would you consider further improving it by putting it into a proper python package/module. Perhaps it would even make most sense as a `Orange.data.Table.from_pandas()` classmethod, with pandas imported in it as needed?

File: pandas/pandas_interoperability [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import Orange
+import pandas as pd
+import numpy as np
+
+
+def construct_domain(df):
+    columns = OrderedDict(df.dtypes)
```
Comment (kernc) [inline pandas/pandas_interoperability]: `OrderedDict` was not imported.

File: pandas/pandas_interoperability [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import Orange
+import pandas as pd
```
Comment (kernc) [inline pandas/pandas_interoperability]: This is a heavy import we'd rather not rely on. Is it possible to import it from within the function providing the functionality. As far as I can tell, importing it is not even needed here. :)

File: pandas/pandas_interoperability [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import numpy as np
+
+
+def construct_domain(df):
+    columns = OrderedDict(df.dtypes)
+
+    def create_variable(col):
+        if col[1].__str__().startswith('float'):
```
Comment (kernc) [inline pandas/pandas_interoperability]: Apparently, one can use `__eq__()` method to test for dtype equality, e.g.
```py
if dtype == float: ...
```

Comment (jamartinh) [inline pandas/pandas_interoperability]: Hi Kern,

I am having an error when using many widgets, I just want to know if this
is could be produced by the way I am converting dataframes between pandas
and Orange.

Best,
Jose

TypeError Traceback (most recent call last):

File "c:\users\jamh\orange3\orange3\Orange\canvas\scheme\widgetsscheme.py",
line 680, in process_signals_for_widget

handler(*args)

File "c:\users\jamh\orange3\orange3\Orange\widgets\utils\sql.py", line 24,
in new_f

return f(self, data, *args, **kwargs)

File "c:\users\jamh\orange3\orange3\Orange\widgets\data\owrank.py", line
256, in setData

self.updateScores()

File "c:\users\jamh\orange3\orange3\Orange\widgets\data\owrank.py", line
313, in updateScores

self.measure_scores[index] = estimator(data)

File "c:\users\jamh\orange3\orange3\Orange\preprocess\score.py", line 48,
in __call__

data = pp(data)

File "c:\users\jamh\orange3\orange3\Orange\preprocess\preprocess.py", line
110, in __call__

attributes = [transform(var) for var in data.domain.attributes]

File "c:\users\jamh\orange3\orange3\Orange\preprocess\preprocess.py", line
110, in <listcomp>

attributes = [transform(var) for var in data.domain.attributes]

File "c:\users\jamh\orange3\orange3\Orange\preprocess\preprocess.py", line
100, in transform

new_var = method(data, var)

File "c:\users\jamh\orange3\orange3\Orange\preprocess\discretize.py", line
121, in __call__

d = distribution.get_distribution(data, attribute)

File "c:\users\jamh\orange3\orange3\Orange\statistics\distribution.py",
line 296, in get_distribution

return Continuous(dat, variable, unknowns)

File "c:\users\jamh\orange3\orange3\Orange\statistics\distribution.py",
line 194, in __new__

return cls.from_data(variable, dat)

File "c:\users\jamh\orange3\orange3\Orange\statistics\distribution.py",
line 213, in from_data

dist, unknowns = data._compute_distributions([variable])[0]

File "c:\users\jamh\orange3\orange3\Orange\data\table.py", line 1308, in
_compute_distributions

unknowns = bn.countnans(m)

File "countnans.pyx", line 107, in countnans.countnans
(bottlechest/src/func/32bit/countnans.c:1788)

File "countnans.pyx", line 698, in countnans.countnans_slow_axisNone
(bottlechest/src/func/32bit/countnans.c:12306)

File "C:\Anaconda3\lib\site-packages\bottlechest\slow\func.py", line 261,
in countnans

return np.sum(np.isnan(arr))

TypeError: ufunc 'isnan' not supported for the input types, and the inputs
could not be safely coerced to any supported types according to the casting
rule ''safe''



2015-10-26 11:05 GMT+01:00 kernc <notifications@github.com>:

> In pandas/pandas_interoperability
> <https://github.com/biolab/orange3/pull/796#discussion_r42975152>:
>
> > @@ -0,0 +1,27 @@
> > +import Orange
> > +import pandas as pd
> > +import numpy as np
> > +
> > +
> > +def construct_domain(df):
> > +    columns = OrderedDict(df.dtypes)
> > +
> > +    def create_variable(col):
> > +        if col[1].__str__().startswith('float'):
>
> Apparently, one can use __eq__() method to test for dtype equality, e.g.
>
> if dtype == float: ...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biolab/orange3/pull/796/files#r42975152>.
>



-- 
/ .- .-.. .-.. / -.-- --- ..- / -. . . -.. / .. ... / .-.. --- ...- .
José Antonio Martín H. (PhD)           E-Mail: jamartinh@fdi.ucm.es
Computer Science Faculty               Phone: (+34) 91 3947650
Complutense University of Madrid       Fax: (+34) 91 3947527
C/ Prof. José García Santesmases,s/n   28040 Madrid, Spain
web: http://www.dacya.ucm.es/jam/
LinkedIn: http://www.linkedin.com/in/jamartinh (Let's connect)
.-.. --- ...- . / .. ... / .- .-.. .-.. / .-- . / -. . . -..

Comment (kernc) [inline pandas/pandas_interoperability]: What is the code that reproduces this?

Comment (jamartinh) [inline pandas/pandas_interoperability]: Hello, this is an error I have. I don't know if it is related.
Simply reading a dataset from a text file. I am running Orange in Windows
10 64 bits python 3.5


[image: Imágenes integradas 1]

2015-12-11 19:42 GMT+01:00 kernc <notifications@github.com>:

> In pandas/pandas_interoperability
> <https://github.com/biolab/orange3/pull/796#discussion_r47388601>:
>
> > @@ -0,0 +1,27 @@
> > +import Orange
> > +import pandas as pd
> > +import numpy as np
> > +
> > +
> > +def construct_domain(df):
> > +    columns = OrderedDict(df.dtypes)
> > +
> > +    def create_variable(col):
> > +        if col[1].__str__().startswith('float'):
>
> What is the code that reproduces this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biolab/orange3/pull/796/files#r47388601>.
>



-- 
/ .- .-.. .-.. / -.-- --- ..- / -. . . -.. / .. ... / .-.. --- ...- .
José Antonio Martín H. (PhD)           E-Mail: jamartinh@fdi.ucm.es
Computer Science Faculty               Phone: (+34) 91 3947650
Complutense University of Madrid       Fax: (+34) 91 3947527
C/ Prof. José García Santesmases,s/n   28040 Madrid, Spain
web: http://www.dacya.ucm.es/jam/
LinkedIn: http://www.linkedin.com/in/jamartinh (Let's connect)
.-.. --- ...- . / .. ... / .- .-.. .-.. / .-- . / -. . . -..

Comment (kernc) [inline pandas/pandas_interoperability]: Which dataset is that? Can you possibly share it?

Comment (jamartinh) [inline pandas/pandas_interoperability]: Thanks kernc, I have it at Home, I will send you a small sample with the
failure.


Best,
Jose



2015-12-14 23:41 GMT+01:00 kernc <notifications@github.com>:

> In pandas/pandas_interoperability
> <https://github.com/biolab/orange3/pull/796#discussion_r47570346>:
>
> > @@ -0,0 +1,27 @@
> > +import Orange
> > +import pandas as pd
> > +import numpy as np
> > +
> > +
> > +def construct_domain(df):
> > +    columns = OrderedDict(df.dtypes)
> > +
> > +    def create_variable(col):
> > +        if col[1].__str__().startswith('float'):
>
> Which dataset is that? Can you possibly share it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biolab/orange3/pull/796/files#r47570346>.
>



-- 
/ .- .-.. .-.. / -.-- --- ..- / -. . . -.. / .. ... / .-.. --- ...- .
José Antonio Martín H. (PhD)           E-Mail: jamartinh@fdi.ucm.es
Computer Science Faculty               Phone: (+34) 91 3947650
Complutense University of Madrid       Fax: (+34) 91 3947527
C/ Prof. José García Santesmases,s/n   28040 Madrid, Spain
web: http://www.dacya.ucm.es/jam/
LinkedIn: http://www.linkedin.com/in/jamartinh (Let's connect)
.-.. --- ...- . / .. ... / .- .-.. .-.. / .-- . / -. . . -..

**Comment by astaric** [comment]
I would like to move this code to a classmethod on Table class and merge this pull request. @jamartinh, if you agree, I can modify this commit and add some tests before merging. Otherwise I kindly ask you to do it :)

**Comment by jamartinh** [comment]
Hi, I will send you the last update of the code.
Things changed from orange 2 to 3.
Best,
Jose

**Comment by jamartinh** [comment]
def pandas_to_orange(df):
    domain, attributes, metas = construct_domain(df)
    orange_table = Orange.data.Table.from_numpy(domain = domain, X =
df[attributes].values, Y = None, metas = df[metas].values, W = None)
    return orange_table


def orange_to_pandas(dt):
    fileIO = StringIO()
    save_csv_IO(dt, fileIO, delimiter = ',')
    pandas_data_frame = pd.read_csv(fileIO)
    del fileIO
    return pandas_data_frame


def construct_domain(df):
    columns = OrderedDict(df.dtypes)
    attributes = OrderedDict()
    metas = OrderedDict()
    for name, dtype in columns.items():

        if issubclass(dtype.type, np.number):
            if len(df[name].unique()) >= 13 or issubclass(dtype.type,
np.inexact):
                attributes[name] = Orange.data.ContinuousVariable(name)
            else:
                df[name] = df[name].astype(str)
                attributes[name] = Orange.data.DiscreteVariable(name,
values = sorted(df[name].unique().tolist()))
        else:
            df[name] = df[name].values.astype(str)
            metas[name] = Orange.data.StringVariable(name)

    domain = Orange.data.Domain(attributes = attributes.values(), metas =
metas.values())

    return domain, list(attributes.keys()), list(metas.keys())


2016-01-24 13:05 GMT+01:00 Jose Antonio Martin H. <xjamartinh@gmail.com>:

> Hi, I will send you the last update of the code.
> Things changed from orange 2 to 3.
> Best,
> Jose
> El 22/1/2016 10:57, "Anže Starič" <notifications@github.com> escribió:
>
>> I would like to move this code to a classmethod on Table class and merge
>> this pull request. @jamartinh <https://github.com/jamartinh>, if you
>> agree, I can modify this commit and add some tests before merging.
>> Otherwise I kindly ask you to do it :)
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/biolab/orange3/pull/796#issuecomment-173865408>.
>>
>


-- 
/ .- .-.. .-.. / -.-- --- ..- / -. . . -.. / .. ... / .-.. --- ...- .
José Antonio Martín H. (PhD)           E-Mail: jamartinh@fdi.ucm.es
Computer Science Faculty               Phone: (+34) 91 3947650
Complutense University of Madrid       Fax: (+34) 91 3947527
C/ Prof. José García Santesmases,s/n   28040 Madrid, Spain
web: http://www.dacya.ucm.es/jam/
LinkedIn: http://www.linkedin.com/in/jamartinh (Let's connect)
.-.. --- ...- . / .. ... / .- .-.. .-.. / .-- . / -. . . -..

**Comment by jamartinh** [comment]
The most up to date code is in:

https://github.com/jamartinh/Orange3-Spark/blob/master/orangecontrib/spark/utils/data_utils.py

2016-01-25 10:13 GMT+01:00 Jose Antonio Martin H. <xjamartinh@gmail.com>:

>
> def pandas_to_orange(df):
>     domain, attributes, metas = construct_domain(df)
>     orange_table = Orange.data.Table.from_numpy(domain = domain, X =
> df[attributes].values, Y = None, metas = df[metas].values, W = None)
>     return orange_table
>
>
> def orange_to_pandas(dt):
>     fileIO = StringIO()
>     save_csv_IO(dt, fileIO, delimiter = ',')
>     pandas_data_frame = pd.read_csv(fileIO)
>     del fileIO
>     return pandas_data_frame
>
>
> def construct_domain(df):
>     columns = OrderedDict(df.dtypes)
>     attributes = OrderedDict()
>     metas = OrderedDict()
>     for name, dtype in columns.items():
>
>         if issubclass(dtype.type, np.number):
>             if len(df[name].unique()) >= 13 or issubclass(dtype.type,
> np.inexact):
>                 attributes[name] = Orange.data.ContinuousVariable(name)
>             else:
>                 df[name] = df[name].astype(str)
>                 attributes[name] = Orange.data.DiscreteVariable(name,
> values = sorted(df[name].unique().tolist()))
>         else:
>             df[name] = df[name].values.astype(str)
>             metas[name] = Orange.data.StringVariable(name)
>
>     domain = Orange.data.Domain(attributes = attributes.values(), metas =
> metas.values())
>
>     return domain, list(attributes.keys()), list(metas.keys())
>
>
> 2016-01-24 13:05 GMT+01:00 Jose Antonio Martin H. <xjamartinh@gmail.com>:
>
>> Hi, I will send you the last update of the code.
>> Things changed from orange 2 to 3.
>> Best,
>> Jose
>> El 22/1/2016 10:57, "Anže Starič" <notifications@github.com> escribió:
>>
>>> I would like to move this code to a classmethod on Table class and merge
>>> this pull request. @jamartinh <https://github.com/jamartinh>, if you
>>> agree, I can modify this commit and add some tests before merging.
>>> Otherwise I kindly ask you to do it :)
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/biolab/orange3/pull/796#issuecomment-173865408>.
>>>
>>
>
>
> --
> / .- .-.. .-.. / -.-- --- ..- / -. . . -.. / .. ... / .-.. --- ...- .
> José Antonio Martín H. (PhD)           E-Mail: jamartinh@fdi.ucm.es
> Computer Science Faculty               Phone: (+34) 91 3947650
> Complutense University of Madrid       Fax: (+34) 91 3947527
> C/ Prof. José García Santesmases,s/n   28040 Madrid, Spain
> web: http://www.dacya.ucm.es/jam/
> LinkedIn: http://www.linkedin.com/in/jamartinh (Let's connect)
> .-.. --- ...- . / .. ... / .- .-.. .-.. / .-- . / -. . . -..
>



-- 
/ .- .-.. .-.. / -.-- --- ..- / -. . . -.. / .. ... / .-.. --- ...- .
José Antonio Martín H. (PhD)           E-Mail: jamartinh@fdi.ucm.es
Computer Science Faculty               Phone: (+34) 91 3947650
Complutense University of Madrid       Fax: (+34) 91 3947527
C/ Prof. José García Santesmases,s/n   28040 Madrid, Spain
web: http://www.dacya.ucm.es/jam/
LinkedIn: http://www.linkedin.com/in/jamartinh (Let's connect)
.-.. --- ...- . / .. ... / .- .-.. .-.. / .-- . / -. . . -..

**Comment by astaric** [comment]
Pandas integration was selected as one of the GSoC 2016 topics. As it will (hopefully) make this PR redundant, I am closing it.

**Pull request closed without merging**